# Introduction

## R Packages

Before starting, you will need the following R packages. 

```{r read_libraries, include=FALSE}
source(fs::path(here::here(),"project_setup","directories.R"))
knitr::read_chunk("project_setup/libraries.R")
```

The following is a list of packages that are used throughout this book that need to be loaded before any analysis:
```{r general_packages_loaded, message = FALSE, warning=FALSE, collapse=FALSE}
#| code-fold: show
```

Additional packages for [Chapter @sec-HIMA] are:
```{r hima_packages, message = FALSE, warning=FALSE}
#| code-fold: show
```

Additional packages for [Chapter @sec-med_lf] are:
```{r med_lf_packages, message = FALSE, warning=FALSE}
#| code-fold: show
```

Additional packages for [Chapter @sec-quasi] are:
```{r lucid_packages, message = FALSE, warning=FALSE}
#| code-fold: show
```

In order to replicate the style of the figures in this book, you will also have to set the ggplot theme:

```{r set_theme, message = FALSE, warning=FALSE}
#| code-fold: show
```



## Extra Functions   

The analyses in this book rely on several custom functions. The details of these extra functions are provided in [Chapter @sec-code]. 

## The Data

```{r read-in load data, include=FALSE}
knitr::read_chunk("project_setup/load_simu_data.R")
```

The data used in this project is based off of simulated data from the Human Early Life Exposome (HELIX) cohort [@RN38]. The data was simulated for one exposure, five omics layers, and one continuous outcome (after publication, this data will be available on github). The format of this data is a named list with 6 elements. It includes separate numeric matrices for each of the 5 omics layers, as well as the exposure and phenotype data. In all datasets in the list, the rows represent individuals and the columns represent omics features. In this analysis, the exposure and outcome are:\

-  **Exposure**: *hs_hg_m_resid*, representing maternal mercury levels\
-  **Outcome**: *ck18_scaled*, representing child liver enzyme levels, a major risk factor for non-alcoholic fatty liver disease (NAFLD).\

You can read this data into R using the following code:\

```{r, collapse=FALSE}
# Read in reduced full scaled data
simulated_data <- readr::read_rds(fs::path(dir_data_hg,
                                       "simulated_HELIX_data.RDS")) 

# Create list of omics data 
omics_name <- c("methylome", "transcriptome","proteome", "miRNA","metabolome")

omics_lst <- simulated_data[which(names(simulated_data) %in% omics_name)]

# readRDS("https://raw.githubusercontent.com/Goodrich-Lab/multiomics_book/blob/main/data/simulate_Hg_ck18_screened_scaled_omics.RDS")
```



### Descriptive Statistics  

@tbl-descript shows the summary statistics for the exposure and phenotype data in this analysis.

```{r data summary}
#| tbl-cap: "Descriptive Statistics for the Simulated Variables"
#| label: tbl-descript
table1::table1(~., data = simulated_data[["phenotype"]])
```

### Mercury exposure and childhood NAFLD risk

In the simulated data,...**hw to update** 

```{r linear regression}
#| fig-cap: "Association between maternal mercury and CK18 in the Simulated Data"
ggplot(data = simulated_data[["phenotype"]], 
       aes(x = hs_hg_m_scaled, y = ck18_scaled)) + 
  geom_point() +
  stat_smooth(method = "lm",
              formula = y ~ x ,
              geom = "smooth") +
  xlab("Maternal Mercury Exposure (Scaled)") +
  ylab("CK-18 Levels (Scaled)")
```


### Correlation of omics features

@fig-cor shows the correlation within and between the omics layers in the simulated data. 

```{r with ComplexHeatmap}
#| label: fig-cor
#| fig-cap: "Heatmap illustrating the correlation of molecular features within and between different omics layers." 
#| fig-height: 6

source(fs::path(dir_proj, "functions", "load_functions.R"))

# Change omics list elements to dataframes
omics_df <- purrr::map(omics_lst, ~as_tibble(.x, rownames = "name")) %>%
  purrr::reduce(left_join, by = "name") %>%
  column_to_rownames("name")

# Correlation Matrix
cormat <- cor(omics_df, method = "pearson")

# Annotations
annotation <- data.frame(
  ftr_name = colnames(cormat),
  Type = get_omic_layer(colnames(cormat)), 
  index = 1:ncol(cormat)) |>
  dplyr::mutate(Type = case_when(
    Type == "Metabolome" ~ "Metabolites", 
    Type == "Transcriptome" ~ "Gene\nExpression", 
    Type == "Methylation" ~ "DNA\nMethylation", 
    TRUE ~ Type))

# Make Plot
Heatmap(cormat, 
        row_split = annotation$Type,
        column_split = annotation$Type,
        show_row_names = FALSE,
        show_column_names = FALSE, 
        heatmap_legend_param = list(title = "Correlation"))
```


